Add introduction to the section on variants and inductive types in the reference manual (#17800)
coq/coq: doc/sphinx/language/core/inductive.rst

Inductive types and recursive functions
=======================================

The :cmd:`Inductive` command allows to define types by case on the

> inhabitants of the type. On the contrary of types defined by the

Reviewer: Should we link to the term definition here as well?
```suggestion
:term:`inhabitants <inhabitant>` of the type. On the contrary of types defined by the
```

:cmd:`Variant` command, constructors can recursively have arguments in

> the type being defined. Inductive types include natural numbers,

Reviewer 2: Misworded?  Variants are not recursive (as stated a few sentences below).

Author: I think, before, there was "On the contrary of types defined by the ..." but I did not realize that the changes that you proposed changed the meaning. Can I let you do a suggestion?

Reviewer 2: How about this?

```suggestion
:term:`inhabitants <inhabitant>` of the type. These constructors can recursively
have arguments in the type being defined.  In contrast, in types defined by the
:cmd:`Variant` command, such recursive references are not permitted.
Inductive types include natural numbers,
```

Author: Yes, I like it.

> The :cmd:`Inductive` command allows to define types by case on the
> :term:`inhabitants <inhabitant>` of the type. On the contrary of types defined by the
> command :cmd:`Variant`, constructors can recursively have arguments in
> the type being defined. So, inductive types include natural numbers,
> lists, well-founded trees, etc. Inhabitants of inductive type can
> recursively nest only a finite number of constructors. So, they are
> well-founded. This distinguishes them from :cmd:`CoInductive` types,
> such as streams, whose constructors can be nested infinitely. In Coq,
> :cmd:`Variant` types thus correspond to the common subset of inductive
> and coinductive types that are non-recursive.

Reviewer 2: ```suggestion
The :cmd:`Inductive` command allows defining types by cases on the
:term:`inhabitants <inhabitant>` of the type. In contrast, in types defined by the
:cmd:`Variant` command, constructors can recursively have arguments in
the type being defined. Inductive types include natural numbers,
lists and well-founded trees. Inhabitants of inductive types can
recursively nest only a finite number of constructors. So, they are
well-founded. This distinguishes them from :cmd:`CoInductive` types,
such as streams, whose constructors can be infinitely nested. In Coq,
:cmd:`Variant` types thus correspond to the common subset of inductive
and coinductive types that are non-recursive.
```

Inductive types and recursive functions
=======================================

The :cmd:`Inductive` command allows defining types by cases on the form of the
:term:`inhabitants <inhabitant>` of the type. These constructors can recursively
have arguments in the type being defined.  In contrast, in types defined by the
:cmd:`Variant` command, such recursive references are not permitted.
Inductive types include natural numbers,
lists and well-founded trees. Inhabitants of inductive types can
recursively nest only a finite number of constructors. So, they are
well-founded. This distinguishes them from :cmd:`CoInductive` types,
such as streams, whose constructors can be infinitely nested. In Coq,
:cmd:`Variant` types thus correspond to the common subset of inductive
and coinductive types that are non-recursive.

Due to the recursive structure of inductive types, functions on
inhabitants of an inductive type have in general to be defined

> recursively. This is provided by the :n:`fix` expression and the

Reviewer: ```suggestion
recursively. Support for defining recursive functions is provided by the :n:`fix` expression and the
```

or:

```suggestion
recursively. This is done using the :n:`fix` expression and the
```

Author: Using the proposal of @jfehrle instead.

> Due to the recursive structure of inductive types, functions on
> inhabitants of an inductive type have in general to be defined
> recursively. This is provided by the :n:`fix` expression and the

Reviewer 2: ```suggestion
Due to the recursive structure of inductive types, functions on
inductive types generally must be defined
recursively using the :n:`fix` expression (see :n:`@term_fix`) or the
```

WDYT?

.. _gallina-inductive-definitions:

> The :cmd:`Inductive` command allows to define types by case on the inhabitants of the type. On the contrary of types defined by the command :cmd:`Variant`, constructors can recursively have arguments in the type being defined. So, inductive types include natural numbers, lists, well-founded trees, etc. Inhabitants of inductive type can recursively nest only a finite number of constructors. So, they are well-founded. This distinguishes them from :cmd:`CoInductive` types whose constructors can be nested infinitely. In Coq, :cmd:`Variant` types thus correspond to the common subset of inductive and coinductive types that are non-recursive.

Reviewer 2: ```suggestion
The :cmd:`Inductive` command allows defining types by cases on the inhabitants of the type. In contrast, types defined by the :cmd:`Variant` command, constructors can recursively have arguments in the type being defined. Inductive types include natural numbers, lists and well-founded trees. Inhabitants of inductive type can recursively nest only a finite number of constructors. So, they are well-founded. This distinguishes them from :cmd:`CoInductive` types, whose constructors can be infinitely nested. In Coq, :cmd:`Variant` types thus correspond to the common subset of inductive and coinductive types that are non-recursive.
```

Please split this into multiple lines that are ~80 characters long.

This needs some work.
"define types by case on the inhabitants of the type" doesn't make sense to me unless you mean "inhabitants of another type"
How about defining "well-founded" and adding maybe add it to the glossary?
Also, would you give a concrete example of a coinductive that's infinitely nested?  That's not at all obvious.

IIRC there is not a clear explanation (with example) of the difference between a coinductive and an inductive.  Could you add that?

> .. _gallina-inductive-definitions:

Reviewer: This label should stay right before the corresponding header. This is what makes the build currently fail.
